client/src/api.rs

```diff
@@ -33,6 +33,9 @@ pub struct InstanceEnsureRequest {
 
     // base64 encoded cloud-init ISO
     pub cloud_init_bytes: Option<String>,
+
+    // If provided, register here to post metrics to Oximeter
+    pub metrics: Option<SocketAddr>,
```

Comment:
So in my first pass I tried something like this, passing metric info to propolis-server on the command line because it does seem like that makes the most sense.  Doing this involved merging the stats endpoint into the existing propolis dropshot server, and that resulted in two issues I then had to solve and required a bit of wrestling with Oximeter to get it to let me create a shared dropshot server that both propolis-server could use, and would be available to serve metrics for Oximeter to collect data from.  I worked a bit with Ben Naecker on this, and we ended up thinking it would be easier to detangle the two services from each other.

The first issue resulting from above is that I need the UUID of the instance to register with Oximeter, and I don't have that when the propolis-server first starts (It comes later in the `instance_ensure()` call).  This requirement comes from the way the initial stat collection registration inside of Oximeter works.

The second issue is, in order to pass down the "ProducerRegistry" object to any library (like crucible) I have to clone it (from the propolis-server dropshot endpoint that is created in server/src/main.rs).  Once I have this clone, I need to attach it to something that `instance_ensure()` can access when it runs.  This can probably be part of `server::Context`, but I think I still need the `Arc<Mutex<Option` around it as I believe Context is shared in a way that requires it.

A third minor issue is that Oximeter will be requesting stats from the propolis-server endpoint every X seconds, and I could not figure out how to turn off just the metrics endpoint from the server log, resulting in a bunch of collection messages.

Reply:
How about moving this back to being a command-line/server option but start server the metrics server in ensure?

Reply:
> How about moving this back to being a command-line/server option but start server the metrics server in ensure?

The metric server does start inside instance_ensure(), but that start only happens if the `InstanceEnsureRequest` `metrics` field contains the metric endpoint (nexus internal API).  I should have mentioned that requirement above.

The three things I need to start the metrics endpoint are:
1. The UUID for the instance.    <-- Comes from Nexus in `api::InstanceEnsureRequest`
2. The I.P. address of the propolis instance.       <-- Comes from server/src/main.rs command line option
3. The I.P. address and port for Nexus (so I can register).   <-- Also from Nexus, though I could potentially use DNS

I can make an option to the server when starting that could request that a metric endpoint will be later started in `instance_insure()`.  To do that I would make `propolis_addr` an`Option` and only start metrics if both `propolis_addr.is_some() && request.metrics.is_some()`.

This would mean to enable metrics, propolis server would have to be started with the metrics option, and the InstanceEnsureRequest has the metric registration.  

Does that sound like the right way to do this?  I'm not sure if wanting a command line option for the server to enable metrics was what you  wanted, or if it was just an idea.

Reply:
Right, (1) makes sense as part of the ensure request and no argument with (2). I meant having (3) passed to propolis on startup because that's not really a property of the instance per-se.

Reply:
> Right, (1) makes sense as part of the ensure request and no argument with (2). I meant having (3) passed to propolis on startup because that's not really a property of the instance per-se.

Ah, yeah, Okay. Thanks for breaking it down.

So,
On startup of propolis-server if I find a metric_register address arg, I will include that metric_register address and the propolis_address as part of the `server::Context` that is created.

Then, in `instance_ensure()`, the presence of metric_register and propolis_address will indicate that it is time to spin up the dropshot endpoint for metrics and, along with the instance UUID I have all the info I need.

I believe this means I can take the `metrics` field out of `propolis_client::api::InstanceEnsureRequest` as that is no longer required, as the decision for metrics has already been made.

This does sound like an improvement, and I think it also satisfies the original concern brought up by @gjcolombo that started this whole thing!  Yeah!

Reply:
I've implemented this idea, so this should be addressed now in the latest commit.